keras/src/backend/jax/trainer.py

```diff
@@ -967,21 +968,35 @@ def _get_jax_state(
         return tuple(state)
 
 
-def _distribute_data(data):
-    distribution = distribution_lib.distribution()
-    if distribution is not None:
+def _distribute_data(data, layouts=None):
+    if layouts is not None:
 
-        def distribute_single_value(d):
-            layout = distribution.get_data_layout(d.shape)
+        def distribute_single_value(d, layout):
             return jax_distribution_lib.distribute_data_input(d, layout)
 
-        return tree.map_structure(distribute_single_value, data)
+        return tree.map_structure(distribute_single_value, data, layouts)
     else:
         return tree.map_structure(jax.device_put, data)
 
 
 class JAXEpochIterator(EpochIterator):
     def _get_iterator(self):
+        layouts = None
+        distribution = distribution_lib.distribution()
+        if isinstance(self.data_adapter, TFDatasetAdapter):
```

Comment:
I think the criteria should be `if distribution is not None:`, you're allowed to distribute from other data adapters.
And conversely, if it is a `TFDatasetAdapter` and distribution is *not* on, you want to do the prefetch iterator, not just `device_put`.

Reply:
Done.